src/hotspot/share/c1/c1_LIRGenerator.cpp

```diff
@@ -1311,6 +1311,35 @@ void LIRGenerator::do_isPrimitive(Intrinsic* x) {
   __ cmove(lir_cond_notEqual, LIR_OprFact::intConst(0), LIR_OprFact::intConst(1), result, T_BOOLEAN);
 }
 
+// Example: Foo.class.getModifiers()
+void LIRGenerator::do_getModifiers(Intrinsic* x) {
+  assert(x->number_of_arguments() == 1, "wrong type");
+
+  LIRItem receiver(x->argument_at(0), this);
+  receiver.load_item();
+  LIR_Opr result = rlock_result(x);
+
+  CodeEmitInfo* info = NULL;
+  if (x->needs_null_check()) {
+    info = state_for(x);
+  }
+
+  LabelObj* L_not_prim = new LabelObj();
+  LabelObj* L_done = new LabelObj();
+
+  LIR_Opr temp = new_register(T_METADATA);
+  // Checking if it's a java mirror of primitive type
+  __ move(new LIR_Address(receiver.result(), java_lang_Class::klass_offset(), T_ADDRESS), temp, info);
+  __ cmp(lir_cond_notEqual, temp, LIR_OprFact::metadataConst(0));
```

Comment:
The previous version of your patch was missing the check for a primitive type. Why did testing not catch this?

Also, your PR title is misleading. You are not simply canonicalizing `Class.getModifier.` You are intrinsifying it.

Reply:
Because is_intrinsic_support does not know the fact that getModifier is intrinsic, so GraphBuilder will not generate Intrinsic instruction, thus canonicalizer is even not applied.

>  Also, your PR title is misleading. You are not simply canonicalizing Class.getModifier. You are intrinsifying it.

Kind remind, it’s indeed mislead  now.

Reply:
Because is_intrinsic_support is not aware of the fact that getModifiers is an intrinsic method. So canonicalizer is even not applied.

> Also, your PR title is misleading. You are not simply canonicalizing Class.getModifier. You are intrinsifying it.

Yes, exactly. Changed.

Reply:
With the new IR test framework, these kinds of problem would never happen, we can do verification at IR level.

Reply:
> Because is_intrinsic_support is not aware of the fact that getModifiers is an intrinsic method. So canonicalizer is even not applied.

But your first version (https://github.com/openjdk/jdk/pull/3616/commits/7f0be4e8a2cb4f6a9ef650b1b31c16f46716a257) already had the change to `is_intrinsic_supported`? Shouldn't that version have triggered test failures because it does not handle primitive types?

> With the new IR test framework, these kinds of problem would never happen, we can do verification at IR level.

Unfortunately not. The IR test framework currently only supports C2.

Reply:
> But your first version (7f0be4e) already had the change to is_intrinsic_supported? Shouldn't that version have triggered test failures because it does not handle primitive types?

The first version only changes Canonicalizer::do_Intrinsic :_

Reply:
Okay, maybe the force-push screwed up some of the history here, you should try to avoid that (it also makes it much harder to review changes incrementally).

Reply:
> Okay, maybe the force-push screwed up some of the history here, you should try to avoid that (it also makes it much harder to review changes incrementally).

Got it! (The reason why I forcibly push is that I can not push them normally after merging some commits from master branch.)

Reply:
Even when merging changes from master into your branch, a force-push should not be required. Here's a related discussion:
https://mail.openjdk.java.net/pipermail/hotspot-dev/2020-November/046538.html